ice40/icestorm.cmake

```diff
@@ -0,0 +1,117 @@
+  set(PYUTILS_PATH ${symbiflow-arch-defs_SOURCE_DIR}/utils:${symbiflow-arch-defs_SOURCE_DIR}/ice40/utils/fasm_icebox)
+
+  get_target_property_required(PYTHON_PREFIX env CONDA_DIR)
+  set(ICESTORM_PREFIX "PREFIX=${PYTHON_PREFIX}")
+
+  add_conda_package(
+    NAME pkg-config
+    PROVIDES pkg-config
+    )
+  add_conda_package(
+    NAME libftdi
+    NO_EXE
+    )
+  get_target_property(LIBFTDI_TARGET env LIBFTDI_TARGET)
+
+  add_conda_pip(
+    NAME numpy
+    NO_EXE
+    )
+
+  add_thirdparty_package(
+    NAME fasm
+    BUILD_INSTALL_COMMAND "cd ${symbiflow-arch-defs_SOURCE_DIR}/third_party/fasm && ${PYTHON3} setup.py develop"
+    PROVIDES fasm
+    DEPENDS ${PYTHON3} ${PYTHON3_TARGET}
+    )
+
+  get_target_property_required(FASM_TARGET env FASM_TARGET)
+
+  set(FASM2ASC ${symbiflow-arch-defs_SOURCE_DIR}/ice40/utils/fasm_icebox/fasm2asc.py)
+  add_custom_target(
+    fasm2asc_deps
+    DEPENDS numpy ${FASM_TARGET} ${FASM2ASC} ${PYTHON3} ${PYTHON3_TARGET}
+    )
+
+  get_target_property_required(SDF_TIMING_TARGET env SDF_TIMING_TARGET)
+  add_custom_target(
+    ice40_import_timing_deps
+    DEPENDS ${ICE40_IMPORT_TIMING} ${SDF_TIMING_TARGET} ${PYTHON3} ${PYTHON3_TARGET}
+    )
+
+  set_target_properties(
+    ice40_import_timing_deps
+    PROPERTIES ICE40_IMPORT_TIMING ${symbiflow-arch-defs_SOURCE_DIR}/ice40/utils/ice40_import_bel_timing.py
+    )
+
+  get_target_property_required(PKG-CONFIG env PKG-CONFIG)
+  get_target_property_required(PKG-CONFIG_TARGET env PKG-CONFIG_TARGET)
+
+  add_thirdparty_package(
+    NAME icestorm
+    PROVIDES iceprog icebox_hlc2asc icebox_vlog icepack icetime icebox
+    BUILD_INSTALL_COMMAND "make -C ${ICESTORM_SRC} ${ICESTORM_PREFIX} PKG_CONFIG=${PKG-CONFIG} install"
+    DEPENDS ${LIBFTDI_TARGET} ${PKG-CONFIG_TARGET}
+    )
+
+  get_target_property_required(ICEBOX_VLOG env ICEBOX_VLOG)
+  get_target_property_required(ICEPACK env ICEPACK)
+  get_target_property_required(ICETIME env ICETIME)
+  get_target_property_required(ICEBOX_HLC2ASC env ICEBOX_HLC2ASC)
+
+  get_target_property(ICEBOX_VLOG_TARGET env ICEBOX_VLOG_TARGET)
+  get_target_property(ICEPACK_TARGET env ICEPACK_TARGET)
+  get_target_property(ICETIME_TARGET env ICETIME_TARGET)
+  get_target_property(ICEBOX_HLC2ASC_TARGET env ICEBOX_HLC2ASC_TARGET)
+
+  get_target_property_required(ICEBOX env ICEBOX)
+  get_filename_component(ICEBOX_PATH ${ICEBOX} DIRECTORY)
+  set(ICEBOX_SHARE ${ICEBOX_PATH}/../share/icebox CACHE PATH "")
+
+  set(PYPATH_ARG "PYTHONPATH=\${ICEBOX_PATH}:${PYUTILS_PATH}")
+  define_arch(
+    ARCH ice40
+    YOSYS_SCRIPT ${symbiflow-arch-defs_SOURCE_DIR}/ice40/yosys/synth.tcl
+    DEVICE_FULL_TEMPLATE \${DEVICE}-\${PACKAGE}
+    VPR_ARCH_ARGS
+      --clock_modeling route
+      --allow_unrelated_clustering off
+      --target_ext_pin_util 0.7
+      --router_init_wirelength_abort_threshold 2
+      --congested_routing_iteration_threshold 0.8
```

Comment:
@litghost I ended up adding this to get more consistent routing completion.

Reply:
Seems reasonable to me.

Reply:
I have tried to use this fix so that we can avoid using https://github.com/SymbiFlow/vtr-verilog-to-routing/issues/108. `picosoc` was able to route, but it actually took a significant amount of iterations (480 to be precise).
Would it be better to reduce the threshold even more (say 0.4/0.5)?
By looking at all the routing logs the iteration would never be more than 50 (this means that if routing is taking more than 50 iterations it's highly probable it encountered some high congestion).

As far as I understand the `congested_routing` starts at iteration number `max_iterations * congested_routing_iteration_threshold`, with max_iteration set to 500, a  threshold of 0.4 seems reasonable IMO.

Reply:
I actually think we should consider a new option.  When I've seen the router struggling, it is usually on the last couple overused nodes (overused nodes 1-3).  @elmsfu / @acomodi Can you go back and check what the router iteration log looked like?  I'm thinking the new option is something akin to "when overused nodes < 5 and does not converge in X iterations, use congestion algo".

Reply:
I like this idea, definitely a better heuristic. I suggest a new issue be opened for that (especially as it is mostly a VPR feature) and we merge this PR unless there are other open comments.